casper/src/main/scala/coop/rchain/casper/merging/DagMerger.scala

```diff
@@ -3,59 +3,102 @@ package coop.rchain.casper.merging
+import coop.rchain.sdk.dag.merging.data._
+import coop.rchain.sdk.dag.merging.DagMergingLogic
 import coop.rchain.shared.{Log, Stopwatch}
 import scodec.bits.ByteVector
+import coop.rchain.shared.syntax._
+import coop.rchain.rspace.syntax._
 
 object DagMerger {
 
-  def costOptimalRejectionAlg: DeployChainIndex => Long =
-    (r: DeployChainIndex) => r.deploysWithCost.map(_.cost).sum
-
-  /**
-    * Merges blocks finalized with the fringe
-    *
-    * @param fringeDiffBlocks all blocks included from previous fringe to new fringe
-    * @param previousFringeState base state for merging, previous fringe state hash
-    * @param index getter for block index
-    * @param historyRepository history repository instance
-    * @param rejectionCostF rejection cost calculation function
-    * @return a tuple (new fringe state hash, rejected deploys - signatures)
-    */
-  def mergeFringe[F[_]: Concurrent: Log](
-      fringeDiffBlocks: Set[BlockHash],
-      previousFringeState: Blake2b256Hash,
-      index: BlockHash => F[Vector[DeployChainIndex]],
+  private def resolveConflicts[F[_]: Concurrent: Log](
+      tips: Set[BlockHash],
+      finalFringe: Set[BlockHash],
+      finalState: Blake2b256Hash,
+      msgMap: Map[BlockHash, Message[BlockHash, Validator]],
+      acceptedFinally: Set[DeployChainIndex],
+      rejectedFinally: Set[DeployChainIndex],
       historyRepository: RhoHistoryRepository[F],
-      rejectionCostF: DeployChainIndex => Long
-  ): F[(Blake2b256Hash, Set[ByteString])] =
-    for {
-      actualSet <- fringeDiffBlocks.toList.traverse(index).map(_.flatten.toSet)
-      // TODO: support for late blocks (with justifications below previousFringeState)
-      lateSet = Set[DeployChainIndex]()
+      blockIndex: BlockHash => F[BlockIndex],
+      rejectionCost: DeployChainIndex => Long
+  ): F[Merge[Blake2b256Hash, DeployChainIndex]] = {
+    val cScope = conflictScope(tips, finalFringe, msgMap.getUnsafe(_: BlockHash).seen)
+    // lowest fringe required to to be able to resolve conflicts
+    val lFringe = lowestFringe(
+      cScope.map(m => msgMap.getUnsafe(m).fringe),
+      msgMap.getUnsafe(_: BlockHash).height
+    )
+    val fScope     = finalScope(finalFringe, lFringe, msgMap.getUnsafe(_: BlockHash).seen)
+    val mergeScope = cScope ++ fScope
+
+    val prepareMergingIndicesF = mergeScope.toList
+      .traverse(b => blockIndex(b).map(x => b -> x.deployChains.toSet))
+      .map(_.toMap)
 
-      branchesAreConflicting = (as: Set[DeployChainIndex], bs: Set[DeployChainIndex]) =>
-        (as.flatMap(_.deploysWithCost.map(_.id)) intersect bs.flatMap(_.deploysWithCost.map(_.id))).nonEmpty ||
-          EventLogMergingLogic.areConflicting(
-            as.map(_.eventLogIndex).toList.combineAll,
-            bs.map(_.eventLogIndex).toList.combineAll
-          )
-      historyReader <- historyRepository.getHistoryReader(previousFringeState)
+    for {
+      mergeScopeIndices <- prepareMergingIndicesF
```

Comment:
This is where we can detect that node is in bootstrap mode - if indices for some blocks in merge set are not available - merging cannot happen and node has to work more in bootstrap mode (finalize fringes trusting rejections written in blocks) so it creates more indices.